File: 152.maximumProductSubarray/maximumProductSubarray.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
using namespace std;

class Solution{
public:
  int maxProduct(vector<int>& nums) {
    int res = 0;
    if (nums.size() == 0)
      return res;
    int maxHere = nums[0];
    int minHere = nums[0];
    res = nums[0];
    for (int i = 1; i < nums.size(); i++) {
      int prevMax = maxHere;
      maxHere = max (nums[i], max (nums[i] * maxHere, nums[i] * minHere) );
      minHere = min (nums[i], min (nums[i] * prevMax, nums[i] * minHere) );
      res = max (res, maxHere);
    }
    return res;
  }
};

int main()
{
  Solution s;
  int a[4] = {2,3,-2,4};
  vector<int> vec(a, a+4);
  cout << s.maxProduct(vec);
  cout << endl;
  return 0;
}

File: 152.maximumProductSubarray/maximumProductSubarray.java
class Solution {
    public int maxProduct(int[] nums) {
        int res = nums[0];
        int[] g = new int[nums.length]; // max prod
        int[] h = new int[nums.length]; // min prod
        g[0] = h[0] = nums[0];
        for (int i = 1; i < nums.length; i++) {
            int c1 = g[i - 1] * nums[i];
            int c2 = h[i - 1] * nums[i];
            g[i] = Math.max(Math.max(c1, c2), nums[i]);
            h[i] = Math.min(Math.min(c1, c2), nums[i]);
            res = Math.max(res, g[i]);
        }
        return res;
    }
}
